lib/routes/password.js

```diff
@@ -143,56 +145,155 @@ module.exports = function (
         var authPW = Buffer(request.payload.authPW, 'hex')
         var wrapKb = Buffer(request.payload.wrapKb, 'hex')
         var authSalt = crypto.randomBytes(32)
+        var sessionTokenId = request.payload.sessionToken
         var password = new Password(authPW, authSalt, verifierVersion)
-        db.deletePasswordChangeToken(passwordChangeToken)
-          .then(
-            function () {
-              return password.verifyHash()
-                .then(
-                  function (verifyHash) {
-                    return password.wrap(wrapKb)
-                      .then(
-                        function (wrapWrapKb) {
-                          return db.resetAccount(
-                            passwordChangeToken,
-                            {
-                              verifyHash: verifyHash,
-                              authSalt: authSalt,
-                              wrapWrapKb: wrapWrapKb,
-                              verifierVersion: password.version
-                            }
-                          )
-                        }
-                      )
+        var wantsKeys = (request.query.keys === 'true')
+        var account, verifyHash, sessionToken, keyFetchToken, verifiedStatus
+
+        if (!sessionTokenId) {
+          // If no sessionToken, this could be a legacy client
+          // attempting to change password, process request as legacy client
+          changePassword()
+            .then(notifyAccount)
+            .then(
+              function () {
+                return {}
+              }
+            )
+            .done(reply, reply)
+
+        } else {
+
+          getVerificationStatus(sessionTokenId)
+            .then(changePassword)
+            .then(notifyAccount)
+            .then(createSessionToken)
+            .then(createKeyFetchToken)
+            .then(createResponse)
+            .done(reply, reply)
+        }
+
+
+        function getVerificationStatus(sessionTokenId) {
+          return Tokens.SessionToken.fromHex(sessionTokenId)
+            .then(
+              function (tokenData) {
+                return tokenData.tokenId
+              }
+            )
+            .then(
+              function (tokenId) {
+                return db.sessionTokenWithVerificationStatus(Buffer(tokenId, 'hex'))
```

Comment:
That sounds good, forgot that there is not a concept of a `verified` session at this point .